R/pagination.R

```diff
@@ -1073,35 +1081,78 @@ paginate_to_mpfs <- function(obj,
                              indent_size = 2,
                              pg_size_spec = NULL,
                              page_num = default_page_number(),
-                             rep_cols = num_rep_cols(obj),
+                             rep_cols = NULL,
                              col_gap = 2,
                              verbose = FALSE) {
+  if (isTRUE(page_num)) {
+    page_num <- "page {i}/{n}"
+  }
+  checkmate::assert_string(page_num, null.ok = TRUE, min.chars = 1)
+
+  # We can return a list of paginated tables and listings
+  if (.is_list_of_tables_or_listings(obj)) {
+    cur_call <- match.call(expand.dots = FALSE)
+    mpfs <- unlist(
+      lapply(obj, function(obj_i) {
+        cur_call[["obj"]] <- obj_i
+        eval(cur_call, envir = parent.frame(3L))
+      }),
+      recursive = FALSE
+    )
+
+    if (!is.null(page_num)) {
+      extracted_cpp <- max(
+        sapply(mpfs, function(mpf) {
+          pf <- prov_footer(mpf)
+          nchar(pf[length(pf)])
+        })
+      )
+      mpfs <- .modify_footer_for_page_nums(mpfs, page_num, extracted_cpp)
+    }
+
+    return(mpfs)
+  }
+
+  if (!is.null(page_num)) {
+    # Only adding a line for pagination -> lpp - 1 would have worked too
+    prov_footer(obj) <- c(prov_footer(obj), page_num)
+  }
+
   mpf <- matrix_form(obj, TRUE, TRUE, indent_size = indent_size)
 
   # Turning off min_siblings for listings
-  if (.is_listing(mpf)) {
+  if (.is_listing_mf(mpf)) {
     min_siblings <- 0
   }
 
+  # Checking colwidths
   if (is.null(colwidths)) {
     colwidths <- mf_col_widths(mpf) %||% propose_column_widths(mpf)
   } else {
+    cur_ncol <- ncol(mpf)
+    if (!.is_listing_mf(mpf)) {
+      cur_ncol <- cur_ncol + as.numeric(mf_has_rlabels(mpf))
+    }
+    # This check will be done inside paginate_to_mpfs in case of lists
+    if (!is.null(colwidths) && length(colwidths) != cur_ncol) {
+      stop(
+        "non-null colwidths argument must have length ncol(x) (+ 1 if row labels are present and if it is a table) [",
+        cur_ncol, "], got length ", length(colwidths)
+      )
+    }
+
     mf_col_widths(mpf) <- colwidths
   }
-  if (NROW(mf_cinfo(mpf)) == 0) {
-    mpf <- mpf_infer_cinfo(mpf, colwidths, rep_cols)
-  }
 
-  # Page numbers
-  if (isTRUE(page_num)) {
-    page_num <- "page {i}/{n}"
+  if (is.null(rep_cols)) {
+    rep_cols <- num_rep_cols(obj) # obj and not mpf bc keycols are rep_cols if listings (not in mpf)
+    if (is(obj, "MatrixPrintForm") && .is_listing_mf(obj)) {
+      rep_cols <- length(.keycols_from_listing(obj))
+    }
```

Comment:
This was the necessary fix in case of `"MatrixPrintForm"` @edelarua. No one uses matrix_form before exporting but it is needed for testing `formatters`, hence all the difficulties in having the same behavior

Reply:
Ideally we would have getter and setters taking care of these differences (in this case `num_rep_cols`) but the number of lines are those ahah. What do you think?

Reply:
Sounds good! If you think it will only be used for testing you could add getters/setters and just keep them internal for our use